internal/databases/mongo/binary/backup.go

```diff
@@ -0,0 +1,319 @@
+package binary
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/wal-g/tracelog"
+	"github.com/wal-g/wal-g/internal"
+	"github.com/wal-g/wal-g/internal/databases/mongo/models"
+	"github.com/wal-g/wal-g/utility"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+const backupType = "binary"
+
+type BackupService struct {
+	Context       context.Context
+	MongodService *MongodService
+	LocalStorage  *LocalStorage
+	BackupStorage *BackupStorage
+
+	MongodBackupMeta *MongodBackupMeta
+
+	BackupDirectoriesMap map[string]*BackupDirectoryMeta
+	BackupFilesMap       map[string]*BackupFileMeta
+}
+
+func GenerateNewBackupName() string {
+	return backupType + "_" + utility.TimeNowCrossPlatformUTC().Format(utility.BackupTimeFormat)
+}
+
+func CreateBackupService(ctx context.Context, mongodService *MongodService, localStorage *LocalStorage,
+	backupStorage *BackupStorage) (*BackupService, error) {
+
+	return &BackupService{
+		Context:       ctx,
+		MongodService: mongodService,
+		LocalStorage:  localStorage,
+		BackupStorage: backupStorage,
+
+		BackupDirectoriesMap: map[string]*BackupDirectoryMeta{},
+		BackupFilesMap:       map[string]*BackupFileMeta{},
+	}, nil
+}
+
+func (backupService *BackupService) DoBackup(backupName string, permanent bool) (err error) {
+	err = backupService.InitializeMongodBackupMeta(backupName, permanent)
+	if err != nil {
+		return err
+	}
+
+	backupCursor, err := backupService.MongodService.GetBackupCursor()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			closeErr := backupCursor.Close(backupService.Context)
+			tracelog.ErrorLogger.Printf("Unable to close backup cursor: %+v", closeErr)
+		}
+	}()
+
+	var backupCursorMeta *BackupCursorMeta
+
+	for backupCursor.TryNext(backupService.Context) {
+		// metadata is the first record in backup cursor
+		if backupCursorMeta == nil {
+			metadataHolder := struct {
+				Metadata BackupCursorMeta `bson:"metadata"`
+			}{}
+			err = backupCursor.Decode(&metadataHolder)
+			if err != nil {
+				return err
+			}
+
+			backupCursorMeta = &metadataHolder.Metadata
+
+			err = backupService.processBackupMetadata(backupCursorMeta)
+			if err != nil {
+				return err
+			}
+		} else {
+			var backupFile BackupCursorFile
+			err = backupCursor.Decode(&backupFile)
+			if err != nil {
+				return err
+			}
+
+			err = backupService.AppendBackupFile(backupFile)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	closeChannel := make(chan string)
+
+	go func() {
+		ticker := time.NewTicker(time.Minute * 1)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-closeChannel:
```

Comment:
Maybe add a check for the `BackupService` context?

Reply:
Sorry, I don't understand. Can you give me fragment of example?

Reply:
> I mean, what happens if the backupService.Context gets canceled?
then to closeChannel should send "Game Over"?